crates/codegen/schema/src/validation/ast/production.rs

```diff
@@ -0,0 +1,147 @@
+                version_map: VersionMap::new(cst_node, versioning.to_owned()),
+            }),
+            types::production::Production::TriviaParser {
+                name,
+                version_map: versioning,
+            } => Rc::new(Self::TriviaParser {
+                name: Node::new(cst_node.get("name"), name.to_owned()),
+                version_map: VersionMap::new(cst_node, versioning.to_owned()),
+            }),
+            types::production::Production::Parser {
+                name,
+                version_map: versioning,
+            } => Rc::new(Self::Parser {
+                name: Node::new(cst_node.get("name"), name.to_owned()),
+                version_map: VersionMap::new(cst_node, versioning.to_owned()),
+            }),
+            types::production::Production::PrecedenceParser {
+                name,
+                version_map: versioning,
+            } => Rc::new(Self::PrecedenceParser {
+                name: Node::new(cst_node.get("name"), name.to_owned()),
+                version_map: VersionMap::new(cst_node, versioning.to_owned()),
+            }),
+        }
+    }
+
+    pub fn name_node(&self) -> &Node<String> {
+        match self {
+            Self::Scanner { name, .. }
+            | Self::TriviaParser { name, .. }
+            | Self::Parser { name, .. }
+            | Self::PrecedenceParser { name, .. } => name,
+        }
+    }
+
+    pub fn version_map_cst_node(&self) -> cst::NodeRef {
+        match self {
+            Production::Scanner { version_map, .. } => version_map.cst_node.clone(),
+            Production::TriviaParser { version_map, .. } => version_map.cst_node.clone(),
+            Production::Parser { version_map, .. } => version_map.cst_node.clone(),
+            Production::PrecedenceParser { version_map, .. } => version_map.cst_node.clone(),
+        }
+    }
+
+    pub fn versions(&self) -> Option<Vec<&Node<Version>>> {
+        match self {
+            Production::Scanner { version_map, .. } => match version_map.value {
+                VersionMap::Unversioned(_) => None,
+                VersionMap::Versioned(ref map) => Some(map.keys().collect()),
+            },
+            Production::TriviaParser { version_map, .. } => match version_map.value {
+                VersionMap::Unversioned(_) => None,
+                VersionMap::Versioned(ref map) => Some(map.keys().collect()),
+            },
+            Production::Parser { version_map, .. } => match version_map.value {
+                VersionMap::Unversioned(_) => None,
+                VersionMap::Versioned(ref map) => Some(map.keys().collect()),
+            },
+            Production::PrecedenceParser { version_map, .. } => match version_map.value {
+                VersionMap::Unversioned(_) => None,
+                VersionMap::Versioned(ref map) => Some(map.keys().collect()),
+            },
+        }
+    }
+}
+
+pub trait ConcreteAbstractPair {
```

Comment:
Not sure I understand the purpose of `ConcreteAbstractPair`. Is there a specific requirement for production types? Otherwise, why is not implemented for other AST nodes?

Reply:
Ignoring for now since we are refactoring validation code.